### PR TITLE
Fix colors in ice plugin manager to match uiThemes

### PR DESCRIPTION
**Pull request by mSlider**
- added styles in light and darks css for ice plugin manager

### CODE REVIEW

**Comment by jojo535275** [comment]
Hola, i just tested the wip.
but still see the collection manager in dark!
here is picture of icestudio, system info and the freshly installed ligth.css file (apparently it contains the color modifications)
Capture ligth

**Comment by jojo535275** [comment]
Oups, sorry Alex, i understood wrongly! 
Your mods are about the plugin Manager (on left side of workspace) and not the collection manager(on the right side).
I'm always mixing those two :-)
So, yes yours mods are working here!

**Comment by mSlider** [comment]
Hi Jo, 

Yes, the fix is only for de ice pluguin manager, but the plugins have their own css.

I think that every pluguin must have their own independent style. But i don’t know what is the Idea in the Carlos mind when he designed the ice pluguin manager.

Alex,
mSlider.com

**Comment by jojo535275** [comment]
Hi Alex,

Personally i will use icestudio on dark mode!
I will, maybe, only use the light mode shortly for printing my designs on paper ! 
And for printing, i will close the collection manager window!
So _for me_ there is no need to have also the collection manager, (or other plugins) in light mode!

Thanks again!

**Comment by cavearr** [comment]
Hi @mSlider ! the plugin manager permits integrate in several forms multiple funcionalities. In fact in near future a lot of features of Icestudio will be migrate as plugins, for example, the build system with apio will be a plugin .

Plugins with frontend could be two flavours.

External plugins, like the serial terminall, could be an external app that could be use an api to communicate with icestudio and interchange information. This kind of plugins are totally independent (styles, coding....).

Other plugins could be "embeded" in Icestudio, like the current collection manager. This kind of plugins run into icestudio and should have an style like Icestudio. The idea is that the plugin api, probide at this kind of plugins of a set of ui widgets and an interface to permit create new ones. For example, the tree ui widget for the collection manager is a "test" of a general tree ui widget.

I'm working hard in the new version of the core with the new plugin api, well documented and with strong basis to permit us build better together. 

Step by step but without stop!

**Comment by jojo535275** [comment]
Ola Alex,

I am actually playing (with my very limited coding skills) around the same subject trying to find a solution for starting the collection manager at icestudio startup!.

if i understood well, when you close and reopen the collection manager plug-in, the plug-in theme is updated.
So why not doing just that? 
Taking apart the fact that the the loading of the collection manager is a bit slow when we have a lot of collection added in our icestudio!

I do not know if it helps but:

when icestudio is started and we using the debug function ( hit F12 keyboard key):

- closing and reopening the collection manager plug-in gives the following message on the console

Capture3


So maybe having a look in /scripts/controllers/menu.js around line1454  will help to understand what method to use for starting and closing the collection manager

**Comment by jojo535275** [comment]
And when i launch a record with the performance tab of the debug function(F12) i see:

- when closing collection manager, the method  this.terminate  is used ( see line342 of /resources/libs/iceplugmanager.js )

- when opening collection manager, two methods  are used ( see line 321 and 213 of /resources/libs/iceplugmanager.js )

IcePlugManager.run @ iceplugmanager.js:321
IcePlugManager.launchEmbedded @ iceplugmanager.js:213

As you can see on the following picture the work in collectionManager.js starts shortly (50msec) after the call of the two methods

Capture4

**Comment by mSlider** [comment]
Thanks for tue info Jo, i’ll give it a try!

Alex,
mSlider.com

**Comment by cavearr** [comment]
Hi ! sorry for the delay with the explanation, in short i'm  posting here.

**Comment by cavearr** [comment]
@mSlider where files are you modifying to do this?  is for focus the explanation.

**Comment by mSlider** [comment]
Hola Carlos,

He modificado: icepluginmanager.js y collectionmanager.js. (Te paso pantalla on adjuntos). Con esto cada vez que se lanza el collection manager, se carga el css con el uiTheme correcto.

La cosa esta que si se cambia el uiTheme, no se como hace un update desde menu.js.

O si hay otra manera (seguro que si) mas elegante. He visto el eventbus o algo asi pero no logro entenderlo. 




Alex,
mSlider.com

**Comment by cavearr** [comment]
Hola Alex!, vale ahora te preparo una explicación d como hacerlo de la forma correcta y aprovecho a explicaros el sistema de señales que aunque la que hay ahora mismo fue un primer prototipo, es la base de la nueva versión que subiré pronto.

Pásame los pantallazos que no los veo

**Comment by mSlider** [comment]
Hola Carlos,

Te adjunto los pantallazos en un .zip A ver si ahora los puedes ver.

Y mil gracias por la paciencia.

Un abrazo!

Alex,
mSlider.com <http://mslider.com/>

**Comment by cavearr** [comment]
Hola Alex! no llegan, intenta entrar en github y comentar directamente aquí ya que creo que por email no adjunta nada.

**Comment by mSlider** [comment]
A ver desde github si me deja pasarte los pantallazos!

collectionManager_JS_L28
iceplugmanager_JS_L223
iceplugmanager_JS_L271
Menu_JS_L823

**Comment by mSlider** [comment]
Hi Carlos, Jo,

- Finally I figured out how to change the collection manager uiTheme acordingly with icestudio's theme.
- Now all I have to do is to change the collection's blocks list .css too... but i will be straight forward.
- Carlos, likewise I'll be wait for your directions in this issue because I want to make it your way.

Here's a test preview.

CM_themes_test

**Comment by cavearr** [comment]
Hi @mSlider ! could you post your final code here like before?

**Comment by jojo535275** [comment]
Well done Alex, félicitation,

I can't wait to have a look at your code in a pull request !
What do you think, guys, about the idea of starting (by default) the Collection Manager at icestudio startup ?
It can always be closed manually (by the few user which do not use it ) by a single click on his top-right cross

Do not hesitate to write in spanish guys,  I decipher-it well ( but i do not write-it at all ;-) )!

Joaquim

**Comment by mSlider** [comment]
Hi Carlos,

Here I put the changes in the code. (please be critical with the code to try to learn)
icm_theme_tests

Hi Jo, I'think in the near future could be to make some type of "config-plugins-menu/popup" to choose the plugin options you will have to launch at start-up and it's config if that plugin has parameters. Carlos said that in the future the intention is that all will be plugin (I think atom, VScode, and like editors works this way).

Today I try to modify iceCollection Items to make it compatible with dark and light uiThemes.

Have a nice day guys!

Alex (mSlider)

**Comment by cavearr** [comment]
Hi Alex! thanks for all!, during the afternoon, i'm reviewing your code and tell you and Jomo.

Thans for all again team! (both of you) i know i'm going with delayed tasks because i'm changing a lot of things and i don't want that you waste your time in things that will be change a lot.

In few hours, i'm reviewing it for the final countdown!

**Comment by cavearr** [comment]
About the popup and other widgets this is the way, but with calm! it's importante go in the same way to join efforts!. I'm working in the roadmap and features that permit work together!

Each plugin will have each panel config, integrate all of them in one place (like vscode for exampe) but firs i need to finish the first version of plugin api to work well and with a robust model.

It's a great idea that you comment us all of you want to do to view if collision or not with the current development and in this way all efforts adds!

Thanks again Jomo and Alexander!

**Comment by mSlider** [comment]
Hi Carlos,

I agree with you that is important that we join efforts to go in the same way... Don't think we are wasting our time, in my case apart from improve some things, one of my own goals is to learn how the app is working. I'd like to code and get experience with FPGAs, microcontrollers, PLCs, applications, etc... 

I've opened a new pull request with the last changes i've made, if you like (Jo and Carlos, and everyone who will) can continue de discussion here: [https://github.com/FPGAwars/icestudio/pull/562](url)

**Comment by cavearr** [comment]
The pull request is merged! i'm preparing the reply, and i think i'm modify your code to explain it better :)

**Comment by cavearr** [comment]
In the current wip I am defining two buses for clarity, one for data (process events ...) and one for GUI actions (interactions with the dom), for example if you want to do something in the dom from a worker, I could use the GUI bus (now there are some API events but I am defining better in the new version, it is like a jquery but for workers).

For example, use us the gui bus, with iceGUI which is like a "distributed" jQuery.

And to try to review all the concepts, the other important thing is that these embedded plugins are attached to the dom as a shadowDom, because in this way we could develop independent widgets in style and dom of icestudio (probably in the near future we could select widgets if its complement is integrated as shadow or normal).

If you want to develop, for example, the configuration panel or the board editor, we can do it and learn together in the process, okay?

The dashboard editor could be a very interesting exercise because you could learn a lot about the basics of fpgas.

I hope, I am explaining to you in this brief assault concept, ask me everything you need. As I tell you, I am finishing the new core version and generating a well defined developer documentation.

In this concrete function with the collection manager, your solution is quite perfect, but I redefined it in another way (this is another solution, I couldn't do the best, but I think it is the easiest and most scalable, I needed to update some little code that I have developed on the new kernel, which is very unstable now, in a few days, I am taking it out, but I am updating with what is necessary to do this):

-First create a method in the collection manager to change the theme, a simple function that takes the current UI theme as an argument and restarts the plugin, calls init and passes the current setting.

 - Then subscribe to the method, we could make up everything you need, for example 'ui.updateTheme'. Why this name, because we could use this event to alert all system components interested in topic updates to do so.
 
 - At last in menu.js, when the theme changes, post the 'ui.updateTheme' event by passing the new theme and calling 'all'. With this 'publishAt' method we could indicate the id of specific components or, in the short term, we could manage 'groups' of modules.
 
 Hope this explanation starts to clarify some of your types, tomorrow I'll send you another explanation for Jomo to explain how to start search filter in collection manager.
 
You see the new code in my last commit :

https://github.com/FPGAwars/icestudio/commit/f28e7697d3324b53d94b97cba45616da1cbcb0ee

 Ask for everything you need.

**Comment by jojo535275** [comment]
Thanks Carlos, for those explanations, they are a bit hard to follow for me. ;-)
But they are very important for us because, with some effort, it can help us to understand the global structure of icestudio Code.

@ Alex:  Yes , do not hesitate to post question about your Fpga design projects on the google group.
There some talented Fpga gurus there ;-). They ( with the help of this userfriendly icestudio tool ) helped me to start understand FPGAs quite fast and with fun.

Regards

**Comment by mSlider** [comment]
Hi Carlos!

Thank you very much for pointing the fixes in the good direction and for your explanaition. It's like a gold for me!

This afternoon i'll read your code changes and try to analyze the work on the shadow dom and the publish-subscribe method you detailed.

If I have time I'll play with it to get the concepts clears.

Jo, sorry for re-fixing your previous fix in the height of the collection manager. I'll try an idea that comes to my mind and I think it will be a clean solution for that. I'm convinced that rhe height must be 100%, but have to put some footer space or bottom pading in order to see uncutted blocks.

It's awesome this way of working together boys!!!

**Comment by jojo535275** [comment]
No problem Alex, my fix was not the cleanest way of doing thinks! 
But very often, when i found a sufficiently good working solution, i tend to stop searching for a better one ;-). 
I'm a lasy guy :-)
Yes, it is really nice to join our different effort/competencies to get thinks done. It can make the final result more interesting!
